File: github_cpp/16/11.cpp


#include <iostream>

using namespace std;

void printArray(int arr[], int size);

int shellSort(int arr[], int n) {
    
    for (int gap = n/2; gap > 0; gap /= 2) {
        
        
        
        
        for (int i = gap; i < n; i += 1) {
            
            
            int temp = arr[i];

            
            
            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                arr[j] = arr[j - gap];

            
            arr[j] = temp;
        }
    }
    printArray(arr, n);
}

void printArray(int arr[], int size) {
    for (int i=0; i < size; i++)
        cout<<" "<<arr[i];
    cout<<endl;
}

int main(int argc, char const *argv[]) {
    int arr[] = {3,1,5,2,75,7,4,12,35,54,1,256,24,5736453,1,7};
    int n = sizeof(arr) / sizeof(arr[0]);
    cout<<"Shell Sort => ";
    shellSort(arr, n);
    return 0;
}

File: github_code_index/java/13/11.java
public class Solution {
  private static int[] shellSort(int[] ar, int h) {
    for (int i = h; i < ar.length; i += h) {
      for (int j = i; j > 0; j -= h) {
        if (ar[j - h] > ar[j]) {
          int buf = ar[j];
          ar[j] = ar[j - h];
          ar[j - h] = buf;
        } else {
          break;
        }
      }
    }
    return ar;
  }
  
  private static void printArray(int[] ar) {
    for (int i = 0; i < ar.length; i++) {
      System.out.print(ar[i] + " ");
    }
    System.out.println();
  }

  public static void main(String[] args) {
    int[] ar = shellSort( new int[]{ 90, 3, 6, 4, 5, 18, 32, 18, 24, 1 }, 4);
    printArray(ar);
    ar = shellSort(ar, 2);
    printArray(ar);
    ar = shellSort(ar, 1);
    printArray(ar);
  }
}
